src/senselab/text/tasks/emotional_analysis/emotional_analysis.py

```diff
@@ -0,0 +1,79 @@
+"""This module provides functionality for performing emotional analysis on text using a specified model."""
+
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+
+from senselab.text.tasks.emotional_analysis.constants import Emotion
+from senselab.utils.data_structures.device import DeviceType, _select_device_and_dtype
+from senselab.utils.data_structures.model import HFModel
+from senselab.utils.interfaces.base_analysis import BaseAnalysis
+
+
+class EmotionalAnalysis(BaseAnalysis):
+    """A class for performing emotional analysis on text."""
+
+    @classmethod
+    def analyze(
+        cls,
+        pieces_of_text: List[str],
+        device: Optional[DeviceType],
+        model: Optional[HFModel] = None,
+        max_length: int = 512,
+        overlap: int = 128,
+        **kwargs: Union[str, int, float, bool],
+    ) -> List[Dict[str, Any]]:
+        """Perform emotional analysis on a list of text pieces.
+
+        Args:
+            pieces_of_text: List of text strings to analyze.
+            device: The device to use for computation.
+            model: The model to use for emotional analysis.
+            max_length: Maximum length of each chunk for long sequences.
+            overlap: Overlap between chunks for long sequences.
+            **kwargs: Additional keyword arguments.
+
+        Returns:
+            A list of dictionaries, each containing:
+                - scores (Dict[str, float]): Probabilities for each emotion.
+                - dominant_emotion (str): The most likely emotion in the text.
+
+        Raises:
+            ValueError: If the input list is empty or None.
+        """
+        if not pieces_of_text:
+            raise ValueError("Input list is empty or None.")
+
+        if model is None:
+            model = HFModel(path_or_uri="j-hartmann/emotion-english-distilroberta-base", revision="main")
+
+        device, torch_dtype = _select_device_and_dtype(
+            user_preference=device, compatible_devices=[DeviceType.CUDA, DeviceType.CPU]
+        )
+
+        tokenizer = cls._get_tokenizer(model, "emotion-analysis")
+        model_instance = cls._load_model(model, "emotion-analysis", device, torch_dtype)
+
+        # Get emotion labels from the model's config
+        emotion_labels = model_instance.config.id2label
+
+        results: List[Dict[str, Any]] = []
+        for text in pieces_of_text:
```

Comment:
what is the advantage of using the tokenizer and the model instead of working with the huggingface text classification pipeline? https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.TextClassificationPipeline

Reply:
I was mainly using the model directly for processing long texts, but can do the same with a pipeline:

Text → text chunks → pass chunks to the pipeline → average the scores for both labels.  
Text → tokenized chunks → pass chunks to the model → softmax for probabilities → average the scores for both labels.

I'll switch to using the pipeline for simplicity.